airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/parquet/JsonSchemaConverter.java

```diff
@@ -0,0 +1,174 @@
+      .addToSchema(Schema.create(Type.LONG));
+  private static final StandardNameTransformer NAME_TRANSFORMER = new StandardNameTransformer();
+
+  private final Map<String, String> standardizedNames = new HashMap<>();
+
+  public Map<String, String> getStandardizedNames() {
+    return standardizedNames;
+  }
+
+  /**
+   * @return - Avro schema based on the input {@code jsonSchema}.
+   */
+  public Schema getAvroSchema(JsonNode jsonSchema,
+                              String name,
+                              @Nullable String namespace,
+                              boolean appendAirbyteFields) {
+    String stdName = NAME_TRANSFORMER.getIdentifier(name);
+    RecordBuilder<Schema> builder = SchemaBuilder.record(stdName);
+    if (!stdName.equals(name)) {
+      standardizedNames.put(name, stdName);
+      LOGGER.warn("Schema name contains illegal character(s) and is standardized: {} -> {}", name,
+          stdName);
+      builder = builder.doc(
+          String.format("%s%s%s",
+              S3ParquetConstants.DOC_KEY_ORIGINAL_NAME,
+              S3ParquetConstants.DOC_KEY_VALUE_DELIMITER,
+              name));
+    }
+    if (namespace != null) {
+      builder = builder.namespace(namespace);
+    }
+
+    JsonNode properties = jsonSchema.get("properties");
+    List<String> fieldNames = new ArrayList<>(MoreIterators.toList(properties.fieldNames()));
+
+    SchemaBuilder.FieldAssembler<Schema> assembler = builder.fields();
+
+    if (appendAirbyteFields) {
+      assembler = assembler.name(JavaBaseConstants.COLUMN_NAME_AB_ID).type(UUID_SCHEMA).noDefault();
+      assembler = assembler.name(JavaBaseConstants.COLUMN_NAME_EMITTED_AT)
+          .type(TIMESTAMP_MILLIS_SCHEMA).noDefault();
+    }
+
+    for (String fieldName : fieldNames) {
+      String stdFieldName = NAME_TRANSFORMER.getIdentifier(fieldName);
+      JsonNode fieldDefinition = properties.get(fieldName);
+      SchemaBuilder.FieldBuilder<Schema> fieldBuilder = assembler.name(stdFieldName);
+      if (!stdFieldName.equals(fieldName)) {
+        standardizedNames.put(fieldName, stdFieldName);
+        LOGGER.warn("Field name contains illegal character(s) and is standardized: {} -> {}",
+            fieldName, stdFieldName);
+        fieldBuilder = fieldBuilder.doc(String.format("%s%s%s",
+            S3ParquetConstants.DOC_KEY_ORIGINAL_NAME,
+            S3ParquetConstants.DOC_KEY_VALUE_DELIMITER,
+            fieldName));
+      }
+      assembler = fieldBuilder.type(getFieldType(fieldName, fieldDefinition)).withDefault(null);
+    }
+
+    return assembler.endRecord();
+  }
+
+  /**
+   * @param fieldDefinition - Json schema field definition. E.g. { type: "number" }.
+   */
+  Schema getFieldType(String fieldName, JsonNode fieldDefinition) {
+    List<JsonSchemaType> fieldTypes = getTypes(fieldName, fieldDefinition.get("type"));
+    // Currently we assume there are at most two type specifications for each field.
```

Comment:
Since Avro supports union types, why do we have this assumption? 

Separately if this assumption is valid, can you make an issue to validate this assumption in the source acceptance tests?

Reply:
Will look into it.

Reply:
This assumption has been removed.